packages/did/src/DidBatcher/DidBatchBuilder.ts

```diff
@@ -0,0 +1,198 @@
+import { ApiPromise } from '@polkadot/api'
+
+import { BN } from '@polkadot/util'
+import type {
+  DidVerificationKey,
+  IIdentity,
+  KeyRelationship,
+  KeystoreSigner,
+  SubmittableExtrinsic,
+} from '@kiltprotocol/types'
+import { SDKErrors } from '@kiltprotocol/utils'
+
+import { FullDidDetails } from '../DidDetails/FullDidDetails.js'
+import { increaseNonce } from '../DidDetails/FullDidDetails.utils.js'
+
+import { generateDidAuthenticatedTx } from '../Did.chain.js'
+import { getSignatureAlgForKeyType } from '../Did.utils.js'
+
+import { checkExtrinsicInput } from './DidBatchBuilder.utils.js'
+
+type BatchInfo = {
+  keyRelationship: KeyRelationship
+  extrinsics: Extrinsic[]
+}
+
+/**
+ * Type of a closure used to select one of the key candidates for a DID to sign a given batch of extrinsics.
+ */
+export type BatchSigningKeySelection = (
+  batch: Extrinsic[],
+  keys: DidVerificationKey[]
+) => Promise<DidVerificationKey>
+/**
+ * The default signign key selection closure which returns the first key in the list of key candidates.
+ *
+ * @param batch The batch of extrinsics to sign. All extrinsics are guaranteed to require the same signing key relationship.
+ * @param keys The list of key candidates that match the required key relationship for the given extrinsic batch.
+ *
+ * @returns The selected key among the list of key candidates. If the returned key is not in the input list of candidates, the signing operation or the extrinsic submission will most likely fail.
+ */
+export const defaultBatchSigningKeySelectionClosure: BatchSigningKeySelection =
+  (batch: Extrinsic[], keys: DidVerificationKey[]) => Promise.resolve(keys[0])
+
+/**
+ * A builder to batch multiple DID-authorised extrinsics which are to be submitted by the same KILT account.
+ */
+export class DidBatchBuilder {
+  private did: FullDidDetails
+  private api: ApiPromise
+
+  protected batches: BatchInfo[] = []
+  protected isConsumed = false
+
+  /**
+   * Create a new builder with the provided WS connection and [[FullDidDetails]].
+   *
+   * @param api The [[ApiPromise]] wrapping the WS connection to a KILT RPC node.
+   * @param did The [[FullDidDetails]] used to fetch the key material needed to authorise the batched operations.
+   */
+  public constructor(api: ApiPromise, did: FullDidDetails) {
+    this.did = did
+    this.api = api
+  }
+
+  private pushNewExtrinsic(
+    ext: Extrinsic,
+    keyRelationship: KeyRelationship
+  ): void {
+    const lastBatch: BatchInfo | undefined =
+      this.batches[this.batches.length - 1]
+
+    // If there was not previous batch, or the new extrinsic requires a different key, create and add a new batch.
+    if (!lastBatch || lastBatch.keyRelationship !== keyRelationship) {
+      this.batches.push({
+        keyRelationship,
+        extrinsics: [ext],
+      })
+      // Otherwise, add the extrinsic to the last batch
+    } else {
+      this.batches[this.batches.length - 1].extrinsics.push(ext)
+    }
```

Comment:
I think it would just add unneeded complexity, in my opinion. It would make more sense to mark all extrinsic in a provided batch to be executed in parallel, but that is not possible if they are anyway part of a batch.

Reply:
yeah let's not tackle that now in any case